File: research/decomp/ssbu-re/gscripts/DumpEffect.py
# Ghidra headless (Jython): decompile string-xref + seed functions -> effect_decomp.txt
from ghidra.app.decompiler import DecompInterface
from ghidra.util.task import ConsoleTaskMonitor

fm = currentProgram.getFunctionManager()
af = currentProgram.getAddressFactory()
ref = currentProgram.getReferenceManager()
dec = DecompInterface(); dec.openProgram(currentProgram)
mon = ConsoleTaskMonitor()
out = open("effect_decomp.txt","w")
def w(s): out.write(s+"\n")

def a(x): return af.getAddress(hex(x).rstrip("L").replace("0x",""))
def func_at(addr): return fm.getFunctionContaining(addr)
def decompile(f):
    if f is None: return "<no function>"
    r = dec.decompileFunction(f, 90, mon)
    if r and r.decompileCompleted(): return r.getDecompiledFunction().getC()
    return "<decompile failed: %s>" % (f.getName())

string_addrs = [0x35832b6,0x35832d8,0x35832fa,0x358331c,0x3583511]
seen=set()
w("===== STRING-XREF FUNCTIONS =====")
for sa in string_addrs:
    for r in ref.getReferencesTo(a(sa)):
        fn=func_at(r.getFromAddress())
        if fn and fn.getEntryPoint().getOffset() not in seen:
            seen.add(fn.getEntryPoint().getOffset())
            w("\n--- %s @ %s ---" % (fn.getName(), fn.getEntryPoint()))
            # also list callers so we can walk up to per-frame calc
            callers=set()
            for cr in ref.getReferencesTo(fn.getEntryPoint()):
                cf=func_at(cr.getFromAddress())
                if cf: callers.add(str(cf.getEntryPoint()))
            w("CALLERS: "+", ".join(sorted(callers)))
            w(decompile(fn))
w("\n===== SEED FUNCTIONS =====")
for s in [0x1eeb0,0x20adc]:
    fn=func_at(a(s))
    if fn:
        w("\n--- seed %s -> %s @ %s ---" % (hex(s), fn.getName(), fn.getEntryPoint()))
        w(decompile(fn))
out.close()
print("DumpEffect: wrote effect_decomp.txt")

File: research/decomp/ssbu-re/gscripts/DumpEffect.java
import ghidra.app.script.GhidraScript;
import ghidra.app.decompiler.*;
import ghidra.program.model.address.*;
import ghidra.program.model.listing.*;
import ghidra.program.model.symbol.*;
import java.io.*;
import java.util.*;

public class DumpEffect extends GhidraScript {
    public void run() throws Exception {
        FunctionManager fm = currentProgram.getFunctionManager();
        ReferenceManager ref = currentProgram.getReferenceManager();
        DecompInterface dec = new DecompInterface();
        dec.openProgram(currentProgram);
        PrintWriter out = new PrintWriter(new FileWriter("effect_decomp.txt"));
        long[] strs = {0x35832b6L,0x35832d8L,0x35832faL,0x358331cL,0x3583511L};
        HashSet<Long> seen = new HashSet<>();
        out.println("===== STRING-XREF FUNCTIONS =====");
        for (long s : strs) {
            Address sa = toAddr(s);
            ReferenceIterator it = ref.getReferencesTo(sa);
            while (it.hasNext()) {
                Reference r = it.next();
                Function f = fm.getFunctionContaining(r.getFromAddress());
                if (f == null) continue;
                long ep = f.getEntryPoint().getOffset();
                if (!seen.add(ep)) continue;
                out.println("\n--- " + f.getName() + " @ " + f.getEntryPoint() + " ---");
                TreeSet<String> callers = new TreeSet<>();
                ReferenceIterator cit = ref.getReferencesTo(f.getEntryPoint());
                while (cit.hasNext()) {
                    Function cf = fm.getFunctionContaining(cit.next().getFromAddress());
                    if (cf != null) callers.add(cf.getEntryPoint().toString());
                }
                out.println("CALLERS: " + callers);
                out.println(decompile(dec, f));
            }
        }
        out.println("\n===== SEED FUNCTIONS =====");
        for (long s : new long[]{0x1eeb0L,0x20adcL}) {
            Function f = fm.getFunctionContaining(toAddr(s));
            if (f != null) {
                out.println("\n--- seed " + Long.toHexString(s) + " -> " + f.getName() + " @ " + f.getEntryPoint() + " ---");
                out.println(decompile(dec, f));
            }
        }
        out.close();
        println("DumpEffect: wrote effect_decomp.txt");
    }
    String decompile(DecompInterface dec, Function f) {
        DecompileResults r = dec.decompileFunction(f, 90, monitor);
        if (r != null && r.decompileCompleted()) return r.getDecompiledFunction().getC();
        return "<decompile failed>";
    }
}
